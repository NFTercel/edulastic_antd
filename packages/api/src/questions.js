import API from './utils/API';

const api = new API();
const prefix = '/question';

const create = data =>
  api
    .callApi({
      url: prefix,
      method: 'post',
      data
    })
    .then(result => result.data.result);

const updateById = (id, data) => {
  delete data._id;
  return api
    .callApi({
      url: `${prefix}/${id}`,
      method: 'put',
      data
    })
    .then(result => result.data.result);
};

const getAll = () =>
  api
    .callApi({
      url: prefix,
      method: 'get'
    })
    .then(result => result.data.result);

const getById = id =>
  api
    .callApi({
      url: `${prefix}/${id}`,
      method: 'get'
    })
    .then(result => result.data.result);

export default {
  create,
  updateById,
  getAll,
  getById
};
