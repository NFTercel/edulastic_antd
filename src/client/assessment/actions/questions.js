import { createAction } from "redux-starter-kit";
import { LOAD_QUESTIONS, GOTO_QUESTION, AUTHOR_QUESTION, ADD_ANSWER } from "../constants/actions";

export const loadQuestionsAction = createAction(LOAD_QUESTIONS);

export const gotoQuestion = question => ({
  type: GOTO_QUESTION,
  payload: {
    question
  }
});

export const addQuestion = question => ({
  type: AUTHOR_QUESTION,
  payload: {
    ...question
  }
});

export const addAnswer = (qid, answer) => ({
  type: ADD_ANSWER,
  payload: {
    qid,
    answer
  }
});
