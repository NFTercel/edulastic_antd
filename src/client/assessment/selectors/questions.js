import { createSelector } from "reselect";
import { values } from "lodash";

const getQuestionsSelector = state => state.assessmentplayerQuestions;

export const getQuestionsListSelector = createSelector(
  getQuestionsSelector,
  state => values(state.byId)
);

export const getQuestionsByIdSelector = createSelector(
  getQuestionsSelector,
  state => state.byId
);
