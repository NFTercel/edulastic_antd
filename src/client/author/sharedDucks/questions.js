import { createAction, createReducer } from "redux-starter-kit";
import { createSelector } from "reselect";
import { values as _values, groupBy as _groupBy, intersection as _intersection, cloneDeep as _cloneDeep } from "lodash";

// actions types
export const LOAD_QUESTIONS = "[author questions] load questions";
export const UPDATE_QUESTION = "[author questions] update questions";
export const SET_FIRST_MOUNT = "[author questions] set first mount";
export const CHANGE_ITEM = "[author questions] change item";
export const CHANGE_ITEM_UI_STYLE = "[author questions] change item ui_style";
export const ADD_QUESTION = "[author questions] add question";
export const CHANGE_CURRENT_QUESTION = "[author quesitons] change current question";
export const ADD_ALIGNMENT = "[author questions] add alignment";
export const REMOVE_ALIGNMENT = "[author questions] remove alignment";

// actions creators
export const loadQuestionsAction = createAction(LOAD_QUESTIONS);
export const addQuestionAction = createAction(ADD_QUESTION);
export const updateQuestionAction = createAction(UPDATE_QUESTION);
export const changeItemAction = createAction(CHANGE_ITEM);
export const changeUIStyleAction = createAction(CHANGE_ITEM_UI_STYLE);
export const setFirstMountAction = createAction(SET_FIRST_MOUNT);
export const changeCurrentQuestionAction = createAction(CHANGE_CURRENT_QUESTION);
export const addAlignmentAction = createAction(ADD_ALIGNMENT);
export const removeAlignmentAction = createAction(REMOVE_ALIGNMENT);

// initialState
const initialState = {
  byId: {},
  current: ""
};

// load questions to the store.
const loadQuestions = (state, { payload }) => {
  state.byId = payload;
};

// update question by id
const updateQuestion = (state, { payload }) => {
  state.byId[payload.id] = payload;
};

const changeItem = (state, { payload }) => {
  const newItem = _cloneDeep(state.byId[state.current]);
  newItem[payload.prop] = payload.value;
  state.byId[state.current] = newItem;
};

const changeUIStyle = (state, { payload }) => {
  const newItem = _cloneDeep(state.byId[state.current]);

  if (!newItem.ui_style) {
    newItem.ui_style = {};
  }

  newItem.ui_style[payload.prop] = payload.value;
  state.byId[state.current] = newItem;
};

const setFirstMount = (state, { id }) => {
  state.byId[id].firstMount = false;
};

// add a new question
const addQuestion = (state, { payload }) => {
  state.byId[payload.id] = payload;
  state.current = payload.id;
};

// change current question
const changeCurrent = (state, { payload }) => {
  state.current = payload;
};

// add alignment to question
const addAlignment = (state, { payload }) => {
  const currentQuestion = state.byId[state.current];

  if (!currentQuestion.alignment || currentQuestion.alignment.length === 0) {
    state.byId[currentQuestion.id].alignment = [payload];
    return;
  }

  let existing = false;

  for (const alignment of currentQuestion.alignment) {
    if (alignment.curriculumId === payload.curriculumId) {
      existing = true;
      const domainGrouped = _groupBy(payload.domain, "id");
      for (const domain of alignment.domain) {
        if (domainGrouped[domain.id]) {
          const selected = domainGrouped[domain.id];
          domains.standards = _intersection(domain.standards, selected.standards, "id");
        }
      }
    }
  }

  if (!existing) {
    currentQuestion.alignment.push(payload);
  }
};

const removeAlignment = (state, { payload }) => {
  const currentQuestion = state.byId[state.current];
  currentQuestion.alignment = currentQuestion.alignment.filter(item => item.curriculumId !== payload);
};

// reducer
export default createReducer(initialState, {
  [LOAD_QUESTIONS]: loadQuestions,
  [UPDATE_QUESTION]: updateQuestion,
  [CHANGE_ITEM]: changeItem,
  [CHANGE_ITEM_UI_STYLE]: changeUIStyle,
  [ADD_QUESTION]: addQuestion,
  [SET_FIRST_MOUNT]: setFirstMount,
  [CHANGE_CURRENT_QUESTION]: changeCurrent,
  [ADD_ALIGNMENT]: addAlignment,
  [REMOVE_ALIGNMENT]: removeAlignment
});

// selectors
const module = "authorQuestions";

export const getCurrentQuestionIdSelector = state => state[module].current;
export const getQuestionsSelector = state => state[module].byId;

// get current Question
export const getCurrentQuestionSelector = createSelector(
  getQuestionsSelector,
  getCurrentQuestionIdSelector,
  (questions, currentId) => questions[currentId]
);

export const getQuestionsArraySelector = createSelector(
  getQuestionsSelector,
  questions => _values(questions)
);

export const getQuestionByIdSelector = (state, qId) => state[module].byId[qId] || {};

// get alignment of current question
export const getQuestionAlignmentSelector = createSelector(
  getCurrentQuestionSelector,
  question => question.alignment || []
);
