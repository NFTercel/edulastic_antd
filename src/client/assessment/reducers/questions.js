import { createReducer } from "redux-starter-kit";
import { LOAD_QUESTIONS } from "../constants/actions";

const initialState = {
  byId: {}
};

const loadQuestions = (state, { payload }) => {
  state.byId = payload;
};

export default createReducer(initialState, {
  [LOAD_QUESTIONS]: loadQuestions
});
