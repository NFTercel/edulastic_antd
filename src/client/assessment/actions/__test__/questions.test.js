import { loadQuestions, gotoQuestion, addQuestion, addAnswer } from "../questions";

describe("Questions", () => {
  it("load questions should return an action", () => {
    expect(
      loadQuestions({
        questions: []
      })
    ).toMatchSnapshot();
  });

  it("go to questions should return an action", () => {
    expect(gotoQuestion({ question: {} })).toMatchSnapshot();
  });

  it("add questions should return an action", () => {
    expect(addQuestion({ question: {} })).toMatchSnapshot();
  });

  it("add answer should return an action", () => {
    expect(addAnswer({ qid: "5e245t52td186247824t" })).toMatchSnapshot();
  });
});
