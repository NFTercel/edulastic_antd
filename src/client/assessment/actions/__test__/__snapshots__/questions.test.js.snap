// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Questions add answer should return an action 1`] = `
Object {
  "payload": Object {
    "answer": undefined,
    "qid": Object {
      "qid": "5e245t52td186247824t",
    },
  },
  "type": "[questions] add answer",
}
`;

exports[`Questions add questions should return an action 1`] = `
Object {
  "payload": Object {
    "question": Object {},
  },
  "type": "author question",
}
`;

exports[`Questions go to questions should return an action 1`] = `
Object {
  "payload": Object {
    "question": Object {
      "question": Object {},
    },
  },
  "type": "[questions] goto question",
}
`;

exports[`Questions load questions should return an action 1`] = `
Object {
  "payload": Object {
    "questions": Object {
      "questions": Array [],
    },
  },
  "type": "[questions] load questions",
}
`;
